_transcripts/032-rust-1.48-1.49.md

```diff
@@ -0,0 +1,731 @@
+`mem::zeroed` and `mem::uninitialized` will now panic if you ever try to use
+them on a type that contains a type that is not allowed to be uninitialized. So,
+for example, in Rust, it is undefined behavior, and it is just not okay to have
+an invalid value. And what that means is, for example, if you have a `bool`, its
+value must be 0 or 1. You're not allowed to have a `bool` that contains, say,
+the bit pattern for a 2. Even though, like in C, this would be totally fine, in
+Rust it is not. Similarly, if you have a reference, then that reference must not
+be dangling; it must not be null; it must point to a valid object. And so you
+can't use `mem::zeroed` for a reference. You can't use `mem::uninitialized` for
+a `bool`. It is just not okay. And now, those functions will panic if you try to
+use them in that way. And it's interesting, because the documentation for them
+doesn't actually say that they panic. But they do say that it's undefined
+behavior, and panic is one form of valid replacement for something that's
+undefined behavior. So this is like an interesting case where the compiler's
+trying to guard you against undefined behavior even when it's allowed to just be
+entirely unhelpful.
+
+__Ben__: Also a reminder that `mem::uninitialized` is deprecated in favor of the
+`MaybeUninit` type. So if you are using `uninitialized`, hopefully you aren't
+hitting it, but please stop using it.
+
+__Jon__: Yes, please use `MaybeUninit` instead.
+
+__Ben__: Yes. And then one more thing you have to talk about.
+
+__Jon__: Yeah, this one's interesting. So this one led me down another rabbit
+hole. I don't know why. I just really like going down these.
+
+This is an issue that was filed in August of 2015 and has been open ever since,
+until 1.48. It used to be that if you had bounds on an associated type of a
+trait, if you constructed a trait object over that trait, the associated bound
+for the trait was just not checked. So what this meant was, if you wrote a
+function that say, took some type that implements a trait, and you called it
+with a trait object whose associated type didn't implement the bounds of the
+associated type. So for example, more concretely here: Let's say you have a
+trait `Foo` that has an associated type `bar`. And in the definition of `bar`,
+you say that `bar` must implement `Copy`. And then you write some function that
+takes a `Foo`. It takes an instance of the trait. And then you call copy with a
+`dyn Foo` where `bar = T` and it's generic over `T`. Notice here that there's no
+bound that says that the `bar` has to be `Copy` on this function. It just takes
+any `T`. It is allowed to call the function that consumes a `Foo`, and that
+function that consumes a `Foo` does not recheck, it doesn't actually validate
+that the input type it got implements `Copy`. It just assumes that it implements
+`Copy`. And so this meant that ever since 2015, you've been able to take a type
```

Comment:
"And then you call copy" <- not sure how to format that. "copy" isn't something you can literally call, is it?

Or was this meant to say, "and then you call that function (that accepts a dyn Foo argument)"?

Reply:
Yeah, "call that function" is appropriate. I think I intended to say that the name of said function was `copy`, but never actually said it :sweat_smile:

Reply:
I could use square brackets to indicate a paraphrasing or editor's correction. It's a little weird because it's confusingly similar to markdown, but seems to render fine.

```text
- And then you call copy with a
+ And then you call [that function] with a
```